proposals/0569-multiline-strings.rst

```diff
@@ -0,0 +1,538 @@
+
+Third party libraries also provide this functionality with quasiquoters, e.g. ``heredoc`` or libraries that also do interpolation like ``string-interpolate``. But Template Haskell is not great:
+
+* It makes compilation difficult in certain environments (e.g. GHCJS)
+* It can slow down compilation
+* Some people avoid it as much as possible out of principle
+* It's a rather heavyweight tool for a seemingly lightweight syntactic feature
+
+Proposed Change Specification
+-----------------------------
+
+#. Lex ``"""`` as an `additional string delimiter <https://gitlab.haskell.org/ghc/ghc/-/blob/8c0ea25fb4a27d4729aabf73f4c00b912bb0c58d/compiler/GHC/Parser/Lexer.x#L577>`_ when the ``MultilineStrings`` language extension is enabled
+
+#. Post-process the string in the following steps:
+
+   #. If the string starts with a newline character, ignore it
+
+   #. Collapse line continuations
+
+   #. Remove common whitespace prefix in every line
+
+      * Blank lines should not be included in this calculation
+
+   #. Interpret escape sequences (occurs after removing whitespace prefix so that literal ``\n`` characters are not included)
+
+#. After parsing, it becomes indistinguishable to the equivalent single-quoted string (modulo annotations for exact-printing)
+
+Line terminators will match whatever line terminators the user is using in the file.
+
+Examples
+--------
+
+Line continuations
+~~~~~~~~~~~~~~~~~~
+
+Line continuations are collapsed first and not included in the whitespace calculation
+
+::
+
+  s =
+      """
+        a b\
+    \ c d e
+        f g
+      """
+
+  -- equivalent to
+  s' = "a b c d e\nf g\n"
+
+Remove common whitespace
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+::
+
+  s =
+        """
+        a b c
+
+        d e f
+
+      g h i
+        """
+
+  -- equivalent to
+  s' = "  a b c\n\n  d e f\n\ng h i\n  "
+
+After lexing, the initial multiline above is parsed as
+
+::
+
+  -- spaces marked as . for visibility
+  [ "......a.b.c"
+  , ""
+  , "......d.e.f"
+  , ""
+  , "....g.h.i"
+  , "......"
+  ]
+
+The blank lines are excluded from the calculation, and we calculate 4 spaces as the shared whitespace prefix, which are removed from every line.
+
+Note that the whitespace preceding the closing ``"""`` is included. This implies that there will be a trailing newline (see the "Trailing newline" example for more information). This also implies that you could change the leading whitespace on every line simply by moving the closing ``"""`` delimiter (although in practice, it'd be recommended to keep the starting/closing delimiters aligned).
+
+::
+
+  -- "a\nb\nc\n"
+  s1 =
+      """
+      a
+      b
+      c
+      """
+
+  -- "  a\n  b\n  c\n"
+  s2 =
+      """
+      a
+      b
+      c
+    """
+
+  -- "a\nb\nc\n  "
+  s3 =
+    """
+    a
+    b
+    c
+      """
+
+Mixing tabs and spaces
+~~~~~~~~~~~~~~~~~~~~~~
+
+A tab and a space both count as a single whitespace character, so mixing the two is inconsequential. But don't do this.
```

Comment:
This is strange when Haskell indentation rules count a tab as 8 spaces (see https://www.haskell.org/onlinereport/haskell2010/haskellch10.html#x17-17800010.3).

I'd stick to that for consistency with Haskell indentation rules. (These days GHC likes to warn about use of tabs anyways.)

Reply:
Ah I forgot about that. But what if the common whitespace prefix is in the middle of a tab? Is the partial tab converted to spaces?

Reply:
Hmm, annoying corner case. But yes, I'd say so.

Or you could simply forbid tabs in the prefix in the proposal, and see if that goes through. (I wouldn't be surprised if if we'd define Haskell now, the whole layout machinery would forbid tabs completely)